core/dbt/include/global_project/macros/materializations/incremental/helpers.sql

```diff
@@ -1,5 +1,6 @@
 
 {% macro incremental_upsert(tmp_relation, target_relation, unique_key=none, statement_name="main") %}
+    
```

Comment:
@matt-winkler @jtcohen6 do you have context on why the existing "target" columns are used as `dest_columns` below ?
Why not use the "tmp" increment's own columns ?

Sorry, I stumbled on this PR because 1. I was trying to implement schema evolution on incremental runs as well and 2. I had issues with some of my increments not necessarily having all the columns of the existing "target".

So I've used tmp_relation instead in the line below in a custom materialization on my side.

If the reason is unclear, I'll open an issue to have a broader discussion on this topic.

Reply:
Hi @b-luu here are my thoughts on use of target columns for each scenario:

- If we `ignore` or `fail` in the case of a schema change (or leave off the `on_schema_change` entirely), we assume that the UPSERT is targeted towards the existing columns in the target table. Basically the thinking here is that the target schema *should not* be changing under these configuration options.
- If we `append_new_columns` or `sync_all_columns`, the column addition / syncing actions have already been performed by the time we reach this point in the code.
- I'd be interested to see more examples of the issue you describe @b-luu to see if we're missing something here; let us know!

@jtcohen6 Please comment if the above seems off